Add project status to projects list

wielinde opened the pull request:
- [x] Show project status in project list
- [x] Do not show project status if project status is not set
- [x] Provide a filter
- [x] Feature specs
- [x] Add sorting on project status

https://community.openproject.com/wp/31375

wielinde committed:
Add project status column to project list

wielinde committed:
Fix layout of enterprise banner on project list

wielinde committed:
Add status filter to project list

Extending the base filter to allow for left outer joins.

wielinde committed:
Do not show the project status in the project list when status not set

wielinde committed:
Adding specs for project list's project status filter.

HDinger commented on app/views/projects/index.html.erb:
```diff
@@ -65,6 +65,7 @@ See docs/COPYRIGHT.rdoc for more details.
           <tr>
             <%= render partial: 'hierarchy_header' %>
             <%= projects_sort_header_tag('name', caption: Project.human_attribute_name(:name), param: :json) %>
+            <%= projects_sort_header_tag('status', caption: Project.human_attribute_name(:status), param: :json) %>
```
This results in an error when trying to sort by status.
<img width="1202" alt="Bildschirmfoto 2019-10-17 um 09 17 32" src="https://user-images.githubusercontent.com/7457313/66986323-fff1cd00-f0be-11e9-8956-addbf8b91b24.png">

wielinde replied:
Thanks. Added the order capability

HDinger commented on app/views/projects/index.html.erb:
```diff
@@ -100,6 +101,12 @@ See docs/COPYRIGHT.rdoc for more details.
+                  <span class="project-status--bulb -inline <%= project_status_css_class(project.status) %>"></span>
+                  <span class="project-status--name <%= project_status_css_class(project.status) %>"><%= project_status_name(project.status) %></span>
+                <% end %>
+              </td>
```
You need to increase the colspan of the description row [here](https://github.com/opf/openproject/blob/dev/app/views/projects/index.html.erb#L147). Otherwise the border is too short.
<img width="370" alt="Bildschirmfoto 2019-10-17 um 09 15 01" src="https://user-images.githubusercontent.com/7457313/66986538-6676eb00-f0bf-11e9-893c-de7c17cbc960.png">

wielinde replied:
Thanks. Fixed.

HDinger reviewed (commented):
Hi @wielinde 

I'm not the right person to evaluate the code for the filters, but the rest looks pretty good. 👍 I only have a few comments. :)

ulferts commented on app/models/queries/projects/filters/project_status_filter.rb:
```diff
@@ -0,0 +1,76 @@
+        def where
+          if operator == '*'
+            # The where condition should always be true
+            '1 = 1'
```
I believe this should be changed to only return projects that have a status (excluding those where the status is not set). At least this is what e.g. a list cf does. And having a filter that filters for ' all' in its current sense would be ineffectual so why having such a filter at all.

wielinde replied:
Hmm, yeah, I asked a couple of people and got a 50/50 result. So I changed it to what you suggested and keep it consistent.

ulferts commented on app/models/queries/projects/filters/project_status_filter.rb:
```diff
@@ -0,0 +1,76 @@
+        end
+
+        def self.key
+          :code
```
I do not like the filter to use that name externally. It admittedly is a matter of taste. If it is to be renamed, I'd opt for `:status_code`.

wielinde replied:
True. I finally took `:project_status_code`

ulferts commented on app/models/queries/projects/filters/project_status_filter.rb:
```diff
@@ -0,0 +1,76 @@
+        private
+
+        def type_strategy
+          @type_strategy ||= ::Queries::Filters::Strategies::ListOptional.new(self)
```
You should be able to simply delete this by simply defining the filter's type to be `:list_optional`

wielinde replied:
Thanks.

ulferts reviewed (changes requested):
On top of the review @HDinger gave (and I would have missed the sort bug :+1:) I added my remarks.

Additionally but not actually related to the PR: The remove icons are not properly aligned:

![image](https://user-images.githubusercontent.com/617519/66990624-db015800-f0c6-11e9-9cbf-cb608350e43b.png)

wielinde committed:
Add sorting for project status to project list and 

change project status filter operator logic to be consistent with other filters

wielinde committed:
Fix position of X-icon for removing a project filter that has no values

wielinde committed:
Fix colspan for expanded row in project list

wielinde commented:
Thank you all for your review. Great findings! I should all be fixed by now.

ulferts reviewed (approved):
Looking all good now :+1:
